src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -114,6 +125,22 @@
                       "/etc/default/puppetserver on Debian systems.")))))))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Helper Functions
+
+(defn construct-404-error-message
+  [jruby-service]
+  (str "Error: Invalid URL - Puppet Server expects requests that conform to the "
+       "/puppet and /puppet-ca APIs.\n\n"
+       "Note that Puppet 3 agents aren't compatible with this version; if you're "
+       "running Puppet 3, you must either upgrade your agents to match the server "
+       "or point them to a server running Puppet 3.\n\n"
+       "Server Info:\n"
+       "  Puppet Server version: " (version-check/get-version-string "puppet-server") "\n"
+       "  Puppet version: " (:puppet-version (config/get-puppet-config jruby-service)) "\n"
```

Comment:
@fpringvaldsen sorry that I didn't spot this sooner, but the call into JRuby is pretty expensive; it'd be much better to do this once at startup and then pass the value into this function rather than hitting JRuby on every 404.  Would you mind opening a new ticket to capture that?